addons/web/controllers/main.py

```diff
@@ -1538,6 +1540,132 @@ def from_data(self, fields, rows):
+                        models_types[cur_model][tok] = attrs['type']
+                        models_comodels[cur_model][tok] = attrs['relation']
+                    else:
+                        models_types[cur_model][tok] = attrs['type']
+                if tok in models_comodels[cur_model]:
+                    cur_model = models_comodels[cur_model][tok]
+
+        self.models_types = models_types
+        self.models_comodels = models_comodels
+
+    def export_record(self, record, model):
+        """Add a record to the root element
+
+        Add a record as an xml element to the xml root element. Can call
+        itself to add related records. If a record as already been
+        exported, the function only return the xml_id.
+
+        Args:
+            record: A recordset containing one record.
+            model: The model of the record to add.
+
+        Returns:
+            The xml_id of the added or existing record.
+        """
+
+
+        if record.id in self.seen_ids[model]:
+            return self.seen_ids[model][record.id]
+
+        xml_id = record._BaseModel__export_xml_id()
+        start_no_xml_id = '__export__.'
+        if xml_id.startswith(start_no_xml_id):
+            name = record.name_get()
+            if name:
+                xml_id = xml_id[len(start_no_xml_id):]+'_'+re.sub('[^a-zA-Z0-9-]', '_', name[0][1])
+            else:
+                xml_id = xml_id[len(start_no_xml_id):]
+        self.seen_ids[model][record.id] = xml_id
+
+        record_element = etree.Element('record', model=model, id=xml_id)
+
+        types = self.models_types[model]
+        comodels = self.models_comodels[model]
+
+        for field_name in types.keys():
+            field = etree.SubElement(record_element, 'field', name=field_name)
+
+            if types[field_name] in ("one2many", "many2many", "many2one") and len(getattr(record, field_name)):
+                if types[field_name] == 'many2one':
+                    field.set('ref', self.export_record(getattr(record, field_name), comodels[field_name]))
+                else:
+                    field.set('eval', '['+', '.join(
+                        '(4,ref(\''+self.export_record(sub_record, comodels[field_name])+'\'))'
+                        for sub_record in getattr(record, field_name)
+                    )+']')
+            else:
+                field.text = unicode(getattr(record, field_name))
+
+        self.root.append(record_element)
+
+        return xml_id
+
+    def base(self, data, token):
+        params = simplejson.loads(data)
+        model, fields, ids, domain = operator.itemgetter('model', 'fields', 'ids', 'domain')(params)
+
+        if ids:
+            records = request.env[model].browse(ids)
+        else:
+            records = request.env[model].search(domain)
+
+        self.extract_types(map(operator.itemgetter('name'), fields), model)
+
+        self.seen_ids = defaultdict(dict)
+        self.root = etree.Element('odoo')
```

Comment:
Uh… no, just no. Pass these things explicitly to the export_record method if you need them there.